### PR TITLE
Added logging to Customer controller

### LINKED ISSUE #26351: Backoffice operations add-edit-delete on Customer / Addresses are not logged in "Logs"
#### Describe the bug

Backoffice operations add-edit-delete on Customer / Addresses are not tracked in "Logs" 

#### Expected behavior

I expect BO operations on all core objects from employees being logged , this is crucial for an admin to track employee activities. This is already correctly happening for products.

I see this open discussion already => https://github.com/PrestaShop/PrestaShop/issues/24698
but the actual PS 1.7 "logging" situation is unclear to me.

#### Steps to Reproduce

1 Add / edit or delete a product
2. Go in Advanced Parameters > Logs
3. See the operation is being tracked correctly with an informative log :
 

4. Add / edit or delete a customer
5. Go in Advanced Parameters > Logs
6. Be sad since nothing is tracked :(

#### Additional information

* PrestaShop version: 1.7.8.0
* PHP version: 7.3

### PR DESCRIPTION
**Pull request by eopws**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Added logging to add-edit-delete operations on a customer in the BO
| Type?             | improvement
| Category?         | BO
| BC breaks?        | Yes
| Deprecations?     | no
| Fixed ticket?     | Fixes #26351
| How to test?      | Follow instructions in the issue

## BC Break
`AbstractCustomerHandler`'s constructor now requires 2 parameters: `TranslatorInterface $translator` & `LoggerInterface $logger`

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/26366)

### CODE REVIEW

File: src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php
Comment (Progi1984) [inline src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php]:
```suggestion
    private function getLogDataContext(?int $id_customer = null, int $error_code = nulll): array
```

File: src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php
Comment (Progi1984) [inline src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php]:
```suggestion
```

**Review by Progi1984 — COMMENTED** [review]
Thank you for your contribution. Could you look to my comments ?

**Comment by eopws** [comment]
Done. @Progi1984 please review

**Review by Progi1984 — CHANGES_REQUESTED** [review]
@eopws Thank you for your contribution. 

I will not approve your PR and I explain why: I will move these logging in Command Handlers rather than in Controller. Because we can add/edit/delete customers from any location. Finally, you can add tests for checking changes.

**Comment by eopws** [comment]
@Progi1984 Thank you for review.

I agreed that we should move the logging instructions to command handlers.
But I followed the style of logging in the product controller.
So the 

1. https://github.com/PrestaShop/PrestaShop/blob/develop/src/PrestaShopBundle/Controller/Admin/ProductController.php#L1074
2. https://github.com/PrestaShop/PrestaShop/blob/develop/src/PrestaShopBundle/Controller/Admin/ProductController.php#L1100
3. https://github.com/PrestaShop/PrestaShop/blob/develop/src/PrestaShopBundle/Controller/Admin/ProductController.php#L1126
and lots more must be updated too

**Comment by eopws** [comment]
> I will move these logging in Command Handlers rather than in Controller

So should I move the logging instructions to https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Form/IdentifiableObject/DataHandler/CustomerFormDataHandler.php
and https://github.com/PrestaShop/PrestaShop/blob/develop/src/Adapter/Customer/CommandHandler/DeleteCustomerHandler.php

**Comment by Progi1984** [comment]
I think I need @PrestaShop/prestashop-maintainers answers, too ;)

**Comment by atomiix** [comment]
I think it's a good idea to move these loggings to the command handlers. We can put the loggings of `ProductController` in the command handlers in another PR later on ;)

**Comment by Progi1984** [comment]
@eopws Could you improve logging for this usecase ? (And we are opened to another contribution for moving logging call in command handlers)

**Comment by eopws** [comment]
> @eopws Could you improve logging for this usecase ? (And we are opened to another contribution for moving logging call in command handlers)

Yes, I am currently working on it

**Comment by Progi1984** [comment]
@eopws You rocks🤘

**Comment by eopws** [comment]
> I will not approve your PR and I explain why: I will move these logging in Command Handlers rather than in Controller. Because we can add/edit/delete customers from any location. Finally, you can add tests for checking changes.

Changed. @Progi1984 @atomiix please review.

P.S.
I followed the style present in https://github.com/PrestaShop/PrestaShop/blob/develop/classes/Mail.php
So Logging commands in product controller should be moved to https://github.com/PrestaShop/PrestaShop/blob/develop/classes/Product.php respectively

**Comment by atomiix** [comment]
@eopws I'm sorry but you added the loggings in the model instead of the command handler.
They should have been added to `src/Adapter/Customer/CommandHandler/AddCustomerHandler.php`, `src/Adapter/Customer/CommandHandler/DeleteCustomerHandler.php`, `src/Adapter/Customer/CommandHandler/EditCustomerHandler.php` etc ;)

**Comment by eopws** [comment]
> @eopws I'm sorry but you added the loggings in the model instead of the command handler. They should have been added to `src/Adapter/Customer/CommandHandler/AddCustomerHandler.php`, `src/Adapter/Customer/CommandHandler/DeleteCustomerHandler.php`, `src/Adapter/Customer/CommandHandler/EditCustomerHandler.php` etc ;)

@atomiix
Oops, I am sorry for my stupidness. I already fixed that, so please review

File: src/Adapter/Customer/CommandHandler/BulkDeleteCustomerHandler.php
Comment (atomiix) [inline src/Adapter/Customer/CommandHandler/BulkDeleteCustomerHandler.php]: The variable `$customer` is not available in this context. Maybe put this call on the loop so we keep track of all the customer's ids that have been deleted?

Comment (atomiix) [inline src/Adapter/Customer/CommandHandler/BulkDeleteCustomerHandler.php]: The variable `$customer` is not available in this context, maybe put this function call in the loop so we can keep track of all customer ids that have been deleted?

Comment (eopws) [reply src/Adapter/Customer/CommandHandler/BulkDeleteCustomerHandler.php]: done

File: src/Adapter/Customer/CommandHandler/AddCustomerHandler.php
Comment (Progi1984) [inline src/Adapter/Customer/CommandHandler/AddCustomerHandler.php]:
```suggestion
use PrestaShopLogger;
use Context;
```

File: src/Adapter/Customer/CommandHandler/BulkDeleteCustomerHandler.php
Comment (Progi1984) [inline src/Adapter/Customer/CommandHandler/BulkDeleteCustomerHandler.php]:
```suggestion
use PrestaShopLogger;
use Context;
```

File: src/Adapter/Customer/CommandHandler/DeleteCustomerHandler.php
Comment (Progi1984) [inline src/Adapter/Customer/CommandHandler/DeleteCustomerHandler.php]:
```suggestion
use PrestaShopLogger;
use Context;
```

File: src/Adapter/Customer/CommandHandler/EditCustomerHandler.php
Comment (Progi1984) [inline src/Adapter/Customer/CommandHandler/EditCustomerHandler.php]:
```suggestion
use PrestaShopLogger;
use Context;
```

**Review by Progi1984 — DISMISSED** [review]
@eopws Sorry... 

The location of logging is perfect.

But could you use dependency injection :
- for translation, you can use `Symfony\Component\Translation\TranslatorInterface` named `@translator` for services YAML
- for logger, you can use `Psr\Log\LoggerInterface` like in `PrestaShop\PrestaShop\Adapter\Import\Handler\AbstractImportHandler`  with `$this->logger->notice`.

**Comment by eopws** [comment]
@Progi1984

> The location of logging is perfect.
> 
> But could you use dependency injection :
> 
> * for translation, you can use `Symfony\Component\Translation\TranslatorInterface` named `@translator` for services YAML
> * for logger, you can use `Psr\Log\LoggerInterface` like in `PrestaShop\PrestaShop\Adapter\Import\Handler\AbstractImportHandler`  with `$this->logger->notice`.

Fixed. Please review

File: src/Adapter/Customer/CommandHandler/BulkDeleteCustomerHandler.php
Comment (atomiix) [inline src/Adapter/Customer/CommandHandler/BulkDeleteCustomerHandler.php]:
```suggestion
```

File: src/Adapter/Customer/CommandHandler/DeleteCustomerHandler.php
Comment (atomiix) [inline src/Adapter/Customer/CommandHandler/DeleteCustomerHandler.php]:
```suggestion
```

**Comment by eopws** [comment]
Review changes are applied. @atomiix please review

**Review by Progi1984 — COMMENTED** [review]
@eopws Could you fix PHPStan feedbacks ?

**Comment by eopws** [comment]
Hope for now everything is fine and ready for merging

File: src/Adapter/Customer/CommandHandler/AddCustomerHandler.php
Comment (Julievrz) [inline src/Adapter/Customer/CommandHandler/AddCustomerHandler.php]:
```suggestion
                'Admin.Advparameters.Feature'
```

File: src/Adapter/Customer/CommandHandler/BulkDeleteCustomerHandler.php
Comment (Julievrz) [inline src/Adapter/Customer/CommandHandler/BulkDeleteCustomerHandler.php]:
```suggestion
                    'Admin.Advparameters.Feature'
```

File: src/Adapter/Customer/CommandHandler/DeleteCustomerHandler.php
Comment (Julievrz) [inline src/Adapter/Customer/CommandHandler/DeleteCustomerHandler.php]:
```suggestion
                'Admin.Advparameters.Feature'
```

**Review by Julievrz — COMMENTED** [review]
I modified the translation domains. You can have a look at the Content style guide for more information: https://zeroheight.com/80a6c5a61/v/latest/p/738e72-manage-translation-domains

I also find it weird to have the id in parenthesis. At first, I understood that as "2 customers were deleted". Especially since the ID is specified in a dedicated column. I know this is how it works for products, but maybe it could be improved with another issue.

image

**Comment by eopws** [comment]
@Julievrz done. But you didn't mark the customer edit file.
https://github.com/PrestaShop/PrestaShop/pull/26366/files#diff-845475576943a5740faae2e961e8218dc1153930b9399a9d5ca8d5d5323ebbfcR123
However I changed domain there too.

**Comment by Julievrz** [comment]
> @Julievrz done. But you didn't mark the customer edit file. https://github.com/PrestaShop/PrestaShop/pull/26366/files#diff-845475576943a5740faae2e961e8218dc1153930b9399a9d5ca8d5d5323ebbfcR123 However I changed domain there too.

Perfect, thanks @eopws!

**Comment by atomiix** [comment]
@eopws It seems that you have 2 merge commits in your git history, can you rebase your branch to get rid of them, please?
Also, you'll have to run `./vendor/bin/php-cs-fixer fix` to fix the coding style ;)

**Comment by eopws** [comment]
Done?

**Comment by PierreRambaud** [comment]
> Done?

Unfortunately, no ^^ https://github.com/PrestaShop/PrestaShop/runs/4434337944?check_suite_focus=true

**Comment by eopws** [comment]
I am sorry for being off. I've fixed the coding style errors. Someone please review and merge if possible

**Review by PierreRambaud — DISMISSED** [review]
I approve the change, but could you please declare the BC break as Handler now require translator and logger?

**Review by atomiix — DISMISSED** [review]

**Review by kpodemski — DISMISSED** [review]

**Comment by boubkerbribri** [comment]
@Progi1984 Could you please dismiss you review so we can test it ?

**Comment by HanaRebaiQA** [comment]
Hello @eopws 

This PR treats only the Customer part as commented in the related issue https://github.com/PrestaShop/PrestaShop/issues/26351#issuecomment-1001593928

- Logs from **BO** : When Create, Edit and Delete Customer ==> OK :heavy_check_mark: 
https://watch.screencastify.com/v/TyHE7GwpzknuRsDGHMcb
image


- Logs from **FO** : NOK :x: 
    -When Create Customer : Error message is displayed with wrong informations
https://watch.screencastify.com/v/w4yRvawuYmhTUlQaLSU2
image

    -When edit Customer : No logs is added
https://watch.screencastify.com/v/KK1SIqi8HfHOUFjeaplQ
image


Waiting for feedback!

Thanks!

**Comment by eopws** [comment]
> -When Create Customer : Error message is displayed with wrong informations

This happens because of email sending on registration, but the email doesn't exist. Should work on real sites with real emails.
Or with ' Send an email after registration' turned off

**Review by atomiix — DISMISSED** [review]

File: src/Adapter/Customer/CommandHandler/AbstractCustomerHandler.php
Comment (Progi1984) [inline src/Adapter/Customer/CommandHandler/AbstractCustomerHandler.php:56]:
```suggestion
    /**
     * @var string
     */
    protected $objectTypeLabel;
```

**Review by Progi1984 — COMMENTED** [review]
@eopws A last feedback from my side and it's an approval

**Comment by kpodemski** [comment]
Hello @PrestaShop/prestashop-maintainers 

Isn't it more scalable to use something like:
https://github.com/thephpleague/tactician-logger/tree/master/tests

?

**Comment by atomiix** [comment]
@kpodemski yep, could be a good idea! Maybe it's out of scope of this PR and it's a more global change to bring though?

**Review by atomiix — APPROVED** [review]

**Comment by kpodemski** [comment]
@atomiix 

yeah, let's leave this PR, and I'll add an issue to log commands globally, it might be helpful :-)

**Review by Progi1984 — APPROVED** [review]

**Review by HanaRebaiQA — COMMENTED** [review]
Hello @eopws 

I checked again this PR. I found these issues : 

**Issue 1** : No added log when delete customer from GDPR module 

1. Create a customer
2. GO to BO > module manager > GDPR 
3. Click on configure
4. Search for customer 
5. Click on `Remove data`
6. Check that the customer is well deleted : OK
7. Check in the logs page : NOK 

https://user-images.githubusercontent.com/92912932/163146771-8898802c-8bc4-4d43-8cf0-c5ba44806550.mp4

**Issue 2** : An exception is thrown when enable / disable selection from Bulk actions in Customers page (It's a regression caused by this PR)

1. Create customers
2. Click to select customers > Bulk actions is activated
3. Click on `enable selection` : NOK
3. Click on `disable selection` : NOK

image

https://user-images.githubusercontent.com/92912932/163148158-7ad64813-237d-4401-9d4e-b6c74ceb7c60.mp4


Could you please check it ?

Thanks!

**Comment by kpodemski** [comment]
Hello @eopws 

There are some errors reported by our QA team. Could you take a look?

**Comment by matks** [comment]
Hi,

Since we had no news from you for more than 30 days, I will follow [our guidelines](https://devdocs.prestashop.com/1.7/project/maintainers-guide/reviewing-pull-requests/#stale-pull-requests) and I'll close this pull request as it seems stale. Feel free to reopen or open another one if you think it's still relevant.

Thanks!

**Pull request closed without merging**
